go/integration_test/processor/gas_billing_test.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package processor
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/Fantom-foundation/Tosca/go/tosca"
+	"github.com/Fantom-foundation/Tosca/go/tosca/vm"
+	op "github.com/ethereum/go-ethereum/core/vm"
+	"go.uber.org/mock/gomock"
+
+	_ "github.com/Fantom-foundation/Tosca/go/processor/floria" // < registers floria processor for testing
+	_ "github.com/Fantom-foundation/Tosca/go/processor/opera"  // < registers opera processor for testing
+)
+
+func TestProcessor_GasBillingEndToEnd(t *testing.T) {
+	tests := map[string]struct {
+		after   WorldState
+		receipt tosca.Receipt
+		result  tosca.Result
+	}{
+		"success": {
+			after: WorldState{
+				{1}: Account{Balance: tosca.NewValue(524000), Nonce: 5},
+				{2}: Account{Balance: tosca.NewValue(0),
+					Code: tosca.Code{
+						byte(vm.PUSH1), byte(0), // < push 0
+						byte(vm.PUSH1), byte(0), // < push 0
+						byte(op.RETURN),
+					},
+				},
+			},
+			receipt: tosca.Receipt{
+				Success: true,
+				GasUsed: 95200,
+			},
+			result: tosca.Result{
+				GasLeft:   5000,
+				Success:   true,
+				GasRefund: 300,
+			},
+		},
+		"failed": {
+			after: WorldState{
+				{1}: Account{Balance: tosca.NewValue(522500), Nonce: 5},
```

Comment:
The only difference between the two after states in these tests seems to be the balance of account 1. It should reduce the code complexity a lot if you only put the balance in the test case specification and build the rest in the test run function below.